scipy/spatial/_qhull.pyx

```diff
@@ -203,6 +203,15 @@ cdef extern from "qhull_src/src/poly_r.h":
 cdef extern from "qhull_src/src/mem_r.h":
     void qh_memfree(qhT *, void *object, int insize)
 
+cdef extern from "Python.h":
+    ctypedef void* PyThread_type_lock
+
+    PyThread_type_lock* PyThread_allocate_lock() nogil
+    void PyThread_free_lock(PyThread_type_lock* lock) nogil
+    int PyThread_acquire_lock(PyThread_type_lock* lock, int flags) nogil
+    void PyThread_release_lock(PyThread_type_lock* lock) nogil
```

Comment:
It probably makes sense to put these bindings elsewhere (maybe in Cython itself?) since other things outside of qhull probably need locks.

Reply:
Maybe we want some internal code in `scipy/_lib` that exposes a scipy-specific lock which is `PyThread_type_lock` on Python 3.10-3.12 and `PyMutex` (once public) on Python >=3.13?

Something like that can be done as a follow-up PR I think. I imagine that that's the pattern that will be useful in other projects as well. Could be upstreamed to Cython too perhaps, but the lead time on that is a few months probably before we can rely on code in Cython 3.1 or 3.2, so let's consider that separately?

Reply:
All definitions above are declared already in Cython here: https://github.com/cython/cython/blob/master/Cython/Includes/cpython/pythread.pxd